### PR TITLE
Symfony3.*: use ArrayChoiceList instead of deprecated SimpleChoiceList

### PR DESCRIPTION
**Pull request by kindratmakc**
I am targetting this branch, because PR isn't break BC.

Changelog

```markdown
### Fixed
- Fix deprecated `SimpleChoiceList`
```

### CODE REVIEW

File: Form/Type/CategorySelectorType.php [hunk 1]
```diff
@@ -58,7 +64,11 @@ public function configureOptions(OptionsResolver $resolver)
             'context' => null,
             'category' => null,
             'choice_list' => function (Options $opts, $previousValue) use ($that) {
-                return new SimpleChoiceList($that->getChoices($opts));
+                if (class_exists('Symfony\Component\Form\Extension\Core\ChoiceList\SimpleChoiceList')) {    // BC with Symfony 2.*
```
Comment (greg0ire) [inline Form/Type/CategorySelectorType.php]: Rather than using old things while still available, how about using new things as soon as available?

Comment (kindratmakc) [inline Form/Type/CategorySelectorType.php]: it makes sense

**Comment by greg0ire** [comment]
Don't try to merge the solution I suggested with the previous solution : just overwrite it.

1. revert to it, undoing the merge in the process : `git reset --hard 62b57b4`
2. overwrite what is on github : `git push --force`

File: Tests/Form/Type/CategorySelectorTypeTest.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+use Sonata\ClassificationBundle\Form\Type\CategorySelectorType;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class CategorySelectorTypeTest extends \PHPUnit_Framework_TestCase
+{
+    public function testConfigureOptions()
+    {
+        $manager = $this->getMock('Sonata\CoreBundle\Model\ManagerInterface');
+        $categorySelectorType = new CategorySelectorType($manager);
+        $categorySelectorType->configureOptions(new OptionsResolver());
```
Comment (greg0ire) [inline Tests/Form/Type/CategorySelectorTypeTest.php]: A test without assertions looks funny, but it's better to run this piece of code to ensure all the classes in use do exist, so good job!

Comment (kindratmakc) [inline Tests/Form/Type/CategorySelectorTypeTest.php]: thx, it my first PR to open source project)

Comment (greg0ire) [inline Tests/Form/Type/CategorySelectorTypeTest.php]: I noticed that, good job!

Comment (OskarStark) [inline Tests/Form/Type/CategorySelectorTypeTest.php]: couldn't we add an assertion here? @greg0ire

Comment (greg0ire) [inline Tests/Form/Type/CategorySelectorTypeTest.php]: We sure can… do you have one in mind?

**Comment by ZRazor** [comment]
I'm new here, so dont judge, if I'm wrong, but this patch wont work as expected on symfony >=3.0
I applied this solution in my project on symfony 3.1 and noticed that category tree doesnt have any "-" symbols for child categories, actually it is not a tree anymore.
'choice_list' option is deprecated in 2.7 (according to documentation http://symfony.com/doc/2.7/reference/forms/types/choice.html )
and is not listed in 3.0 and above ( http://symfony.com/doc/3.0/reference/forms/types/choice.html )
I debugged a little to find out, why it doesnt throw exception:
Symfony\Component\Form\Extension\Core\Type::buildForm
In 2.7 it goes staight to this line:
```
$builder->addViewTransformer(new ChoiceToValueTransformer($options['choice_list']));
``` 
And in 3.0:
```
        $choiceList = $this->createChoiceList($options);
        $builder->setAttribute('choice_list', $choiceList);
```
And createChoiceList returns choices from choices option or from choice_loader, which is created by default. And then overrides anything we have passed to choice_list.
The solution for symfony 3.0 and above is to use 'choice_loader'. Need to pass there a class implementing ( http://api.symfony.com/3.1/Symfony/Component/Form/ChoiceList/Loader/ChoiceLoaderInterface.html )
In symfony 3.2 there is a new class CallbackChoiceLoader
```
    public function configureOptions(OptionsResolver $resolver)
    {
        $that = $this;
        $resolver->setDefaults(array(
            'context' => null,
            'category' => null,

            'choice_loader' => new CallbackChoiceLoader(function() use ($that) {
                return array_flip($that->getChoices());
            })
        ));
    }

```
Also I had to remove Options param from getChoices and childWalker functions, not sure if this is a good solution, but i checked - it never contains category option.

**Comment by greg0ire** [comment]
Would you like to provide a PR @ZRazor ?

**Comment by ZRazor** [comment]
@greg0ire, of course. I'try to make it compatible with 2.7 and >=3.0

**Comment by greg0ire** [comment]
Sorry, but we won't merge that unless it's compatible with 2.3 :P … until the next major release, but I'm afraid it's not coming very soon.

**Comment by ZRazor** [comment]
@greg0ire, i get it. I'll make sure that older versions work fine (with SimpleChoiceList)

**Comment by greg0ire** [comment]
@ZRazor awesome!

**Comment by OskarStark** [comment]
Closing in favour of: https://github.com/sonata-project/SonataClassificationBundle/pull/280#issuecomment-282038580

**Pull request closed without merging**